### PR TITLE
[GlobalISel] Introduce G_TRAP, G_DEBUGTRAP, G_UBSANTRAP

### PR DESCRIPTION
**Pull request by e-kud**
Here we introduce three new GMIR instructions to cover a set of trap intrinsics. The idea behind it is that generic intrinsics shouldn't be used with G_INTRINSIC opcode.

These new instructions can match perfectly with existing trap ISD nodes. It allows X86, AArch64, RISCV and Mips to reuse SelectionDAG patterns for selection and avoid manual selection. However AMDGPU is an exception. It selects traps during legalization regardless SelectionDAG or GlobalISel.

Since there are not many places where traps are used, this change attempts to clean up all the usages of G_INTRINSIC with trap intrinsics. So, there is no stage when both G_TRAP and G_INTRINSIC_W_SIDE_EFFECTS(@llvm.trap) are allowed.

### CODE REVIEW

**Comment by e-kud** [comment]
The idea of having these opcodes has been suggested by @arsenm in https://github.com/llvm/llvm-project/pull/83498#discussion_r1508565901

I find it useful as X86, AArch64, RISCV and Mips can avoid manual selection. AMDGPU is an exception as it lowers traps during legalization even using Selection DAG.

I'm not sure whether it is a good idea or not to edit all targets at once. But I'd like to avoid intermediate stage when both `G_*TRAP` and `G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.*trap)` are allowed.

I've tried to cleanup all `G_INTRINSIC_W_SIDE_EFFECTS.*trap` usages as well as  `llvm.*trap`.

Let me know if the change is undesired, then I'll simply merge aforementioned fix for `X86`.

File: llvm/docs/GlobalISel/GenericOpcode.rst [hunk 1]
```diff
@@ -779,11 +779,17 @@ G_ATOMIC_CMPXCHG
...
+^^^^^^^^^^^^^^^^^
+
+.. |all_g_atomicrmw| replace:: G_ATOMICRMW_XCHG, G_ATOMICRMW_ADD,
+                               G_ATOMICRMW_SUB, G_ATOMICRMW_AND,
+                               G_ATOMICRMW_NAND, G_ATOMICRMW_OR,
+                               G_ATOMICRMW_XOR, G_ATOMICRMW_MAX,
+                               G_ATOMICRMW_MIN, G_ATOMICRMW_UMAX,
+                               G_ATOMICRMW_UMIN, G_ATOMICRMW_FADD,
+                               G_ATOMICRMW_FSUB, G_ATOMICRMW_FMAX,
+                               G_ATOMICRMW_FMIN
```
Comment (arsenm) [inline llvm/docs/GlobalISel/GenericOpcode.rst:792]: Unrelated change

**Review by arsenm — APPROVED** [review]
lgtm except unrelated doc change

File: llvm/docs/GlobalISel/GenericOpcode.rst (hunk 1 above)
Comment (e-kud) [reply llvm/docs/GlobalISel/GenericOpcode.rst]: Dropped the change. Will create a separate PR for this fix.

**Review by aemerson — APPROVED** [review]

**Review by arsenm — APPROVED** [review]

**Comment by e-kud** [comment]
> Can you put this in the PR description so it will end up in the commit message.

@topperc I've rephrased in more general terms. 

I've noticed that there are checks for GMIR opcodes in MachineVerifier. Should I add sanity checks that `G_TRAP` and `G_DEBUGTRAP` have no args and `G_UBSANTRAP` has only an `i8` immediate argument?

**Comment by arsenm** [comment]
> > Can you put this in the PR description so it will end up in the commit message.
> 
> @topperc I've rephrased in more general terms.
> 
> I've noticed that there are checks for GMIR opcodes in MachineVerifier. Should I add sanity checks that `G_TRAP` and `G_DEBUGTRAP` have no args and `G_UBSANTRAP` has only an `i8` immediate argument?

You should add the tests. I believe the operand number checks should work by default, but you might need code to verify the immediate operand

**Review by arsenm — APPROVED** [review]

**Review by qcolombet — APPROVED** [review]
